src/dns.rs

```diff
@@ -0,0 +1,97 @@
+mod local;
+mod default;
+mod cache;
+
+use crate::types::Result;
+use std::net::IpAddr;
+use std::time;
+
+/// A single IP entry in the DNS cache. Each ip entry has an expiration time (based on the TTL)
+/// and a last-used timestamp. When the entry is expired, the entry is found to be stale and should
+/// be refreshed by the resolver. When expired == 0, the entry is considered to be a local override
+/// and should not be refreshed by the resolver.
+/// @todo: Does that information belong here? Or should it be in the cache layer?
+struct IpEntry {
+    /// Actual IP address (either ipv4 or ipv6)
+    ip: IpAddr,
+    /// Epoch when ip address expires
+    expired: u64,
+    /// Epoch when this entry was last used
+    last_used: u64,
+}
+
+/// A DNS entry is a simple domain name to IP address mapping
+#[derive(Default)]
+struct DNSEntry {
```

Comment:
The DNS entry now consists of a domain name, and 2 lists of ip addresses, the ipv4 andipv6. It now contains the roundrobin pointers as well. The whole round robin is encapsulated in the DNSEntry so we can simply do next_ipv4() and not need to worry about it from the outside world.